pkg/agent/podresapi/client.go

```diff
@@ -0,0 +1,165 @@
+	timeout           = 10 * time.Second
+	maxSize           = 1024 * 1024 * 16
+)
+
+var (
+	errGetDisabled = fmt.Errorf("PodResources API Get method disabled")
+	log            = logger.Get("podresapi")
+)
+
+// WithSocketPath sets the kubelet socket path to connect to.
+func WithSocketPath(path string) ClientOption {
+	return func(c *Client) {
+		c.socketPath = path
+	}
+}
+
+// WithClientConn sets a pre-created gRPC connection for the client.
+func WithClientConn(conn *grpc.ClientConn) ClientOption {
+	return func(c *Client) {
+		c.conn = conn
+	}
+}
+
+// NewClient creates a new Pod Resources API client with the given options.
+func NewClient(options ...ClientOption) (*Client, error) {
+	c := &Client{
+		socketPath: defaultSocketPath,
+	}
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if c.conn == nil {
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+
+		dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+			return (&net.Dialer{}).DialContext(ctx, "unix", addr)
+		}
+
+		conn, err := grpc.DialContext(
+			ctx,
+			c.socketPath,
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+			grpc.WithContextDialer(dialer),
+			grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxSize)),
+			grpc.WithBlock(),
+		)
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect podresource client: %w", err)
+		}
+
+		c.conn = conn
+	}
+
+	c.cli = api.NewPodResourcesListerClient(c.conn)
+	return c, nil
+}
+
+// Close closes the client.
+func (c *Client) Close() {
+	if c != nil && c.conn != nil {
+		c.conn.Close()
+		c.conn = nil
+	}
+	c.cli = nil
+}
+
+// HasClient returns true if the client has a usable client.
+func (c *Client) HasClient() bool {
+	return c != nil && c.cli != nil
+}
+
+// List lists all pods' resources.
+func (c *Client) List(ctx context.Context) (PodResourcesList, error) {
+	if !c.HasClient() {
+		return nil, nil
+	}
+
+	reply, err := c.cli.List(ctx, &api.ListPodResourcesRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pod resources by list: %w", err)
+	}
+
+	return PodResourcesList(reply.GetPodResources()), nil
+}
+
+// Get queries the given pod's resources.
+func (c *Client) Get(ctx context.Context, namespace, pod string) (*PodResources, error) {
+	if !c.HasClient() {
+		return nil, nil
+	}
+
+	if !c.noGet {
+		reply, err := c.cli.Get(ctx, &api.GetPodResourcesRequest{
+			PodNamespace: namespace,
+			PodName:      pod,
+		})
+		if err == nil {
+			return &PodResources{reply.GetPodResources()}, nil
+		}
+
+		if !strings.Contains(fmt.Sprintf("%v", err), fmt.Sprintf("%v", errGetDisabled)) {
+			return nil, fmt.Errorf("failed to get pod resources: %w", err)
+		}
+
+		log.Warnf("PodResources API Get() disabled, falling back to List()...")
+
+		c.noGet = true
+	}
+
+	l, err := c.List(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pod resources: %w", err)
+	}
+
+	return l.GetPodResources(namespace, pod), nil
```

Comment:
should we cache whole reply? not only for single pod if we got the whole list....

Reply:
Yes, I was considering that, but decided to save that for another PR.

Reply:
Also I'm not sure if the current podresapi handling during CreateContainer is guaranteed to be always correct... or if is only guaranteed to be correct for the first creation of a named container within the pod. IOW, I'm not sure if we'd need to refetch the pod resources when an already formerly existing container exits/crashes in a pod and gets recreated...

Reply:
To best of my memory, this data is populated from devices/resources allocated at pod admission time. those are not re-allocated on container restarts. So, in theory, if we got info about the pod, it should be complete. I don't know if kubelet handles properly race conditions if pod is in the middle of admission at the time of List() call. (As admission is pod level thingy, I suspect it shouldn't be an issue).
But I agree, we can leave it as separate PR.